store/sqlstore/group_supplier.go

```diff
@@ -0,0 +1,864 @@
+	if err != nil {
+		if err == sql.ErrNoRows {
+			result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.sql_group.no_rows", nil, "Id="+groupID+", "+err.Error(), http.StatusNotFound)
+		} else {
+			result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.select_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		return result
+	}
+
+	if groupSyncable.DeleteAt != 0 {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.sql_group.group_syncable_already_deleted", nil, "group_id="+groupID+"syncable_id="+syncableID, http.StatusBadRequest)
+		return result
+	}
+
+	time := model.GetMillis()
+	groupSyncable.DeleteAt = time
+	groupSyncable.UpdateAt = time
+
+	var rowsAffected int64
+	switch groupSyncable.Type {
+	case model.GroupSyncableTypeTeam:
+		rowsAffected, err = s.GetMaster().Update(groupSyncableToGroupTeam(groupSyncable))
+	case model.GroupSyncableTypeChannel:
+		rowsAffected, err = s.GetMaster().Update(groupSyncableToGroupChannel(groupSyncable))
+	default:
+		model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "model.group.type.app_error", nil, "group_id="+groupSyncable.GroupId+", syncable_id="+groupSyncable.SyncableId+", "+err.Error(), http.StatusInternalServerError)
+		return result
+	}
+
+	if err != nil {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.update_error", nil, err.Error(), http.StatusInternalServerError)
+		return result
+	}
+
+	if rowsAffected == 0 {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.sql_group.no_rows_affected", nil, "", http.StatusInternalServerError)
+		return result
+	}
+
+	result.Data = groupSyncable
+
+	return result
+}
+
+// PendingAutoAddTeamMembers returns a slice of UserTeamIDPair that need newly created memberships
+// based on the groups configurations.
+//
+// Typically minGroupMembersCreateAt will be the last successful group sync time.
+func (s *SqlSupplier) PendingAutoAddTeamMembers(ctx context.Context, minGroupMembersCreateAt int64, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	query := `
+		SELECT 
+			GroupMembers.UserId, GroupTeams.TeamId
+		FROM 
+			GroupMembers
+			JOIN GroupTeams 
+			ON GroupTeams.GroupId = GroupMembers.GroupId
+			JOIN Groups ON Groups.Id = GroupMembers.GroupId
+			JOIN Teams ON Teams.Id = GroupTeams.TeamId
+			%s JOIN TeamMembers 
+			ON 
+				TeamMembers.TeamId = GroupTeams.TeamId 
+				AND TeamMembers.UserId = GroupMembers.UserId
+		WHERE 
+			TeamMembers.UserId IS NULL
+			AND Groups.DeleteAt = 0
+			AND GroupTeams.DeleteAt = 0
+			AND GroupTeams.AutoAdd = true
+			AND GroupMembers.DeleteAt = 0
+			AND Teams.DeleteAt = 0
+			AND (GroupMembers.CreateAt >= :MinGroupMembersCreateAt
+			OR GroupTeams.UpdateAt >= :MinGroupMembersCreateAt)`
+
+	sql := fmt.Sprintf("%s UNION %s", fmt.Sprintf(query, "LEFT"), fmt.Sprintf(query, "RIGHT"))
```

Comment:
I don't understand the second half of the `UNION` directive here. That query does a `RIGHT JOIN TeamMembers`, but still has the `TeamMembers.UserId IS NULL` condition in the `WHERE` clause, which means the `TeamMembers` row must not have existed (there's no way `UserId IS NULL` there), but then I think we're just back to the first half of the `UNION` query.

Can you clarify?

Reply:
^ ping re: UNION

Reply:
The history of this is that it was written [originally using `FULL JOIN`](https://github.com/mattermost/mattermost-server/commit/9d3703bea763620a45ca8020cd63ff17d3a348a1#diff-0b7dcefeb566578492d6be5ea6b2453c) which isn't supported by MySQL so this is the workaround.

Reply:
`LEFT JOIN ... UNION ... RIGHT JOIN` as an alternative to `FULL JOIN` makes sense to me (in so far as anything MySQL-related makes sense to me ;P), but I'm trying to understand what new rows are intended to be added from the `RIGHT JOIN TeamMembers` given the `TeamMembers.UserId IS NULL` constraint in both subqueries. It seems like that should always be an empty set?

It's altogether possible I'm misreading the query here, so at a minimum I'm hoping we can clarify this and add some inline documentation for my future, forgetful self. Maybe we can hop on a call to clarify this?

Reply:
Take a look at the changed query and tell me if that is clearer to read.

Reply:
Yep! If I'm reading the diff correctly, the `UNION` and `RIGHT JOIN` is now just gone, which makes sense. 👍